language/move-lang/stdlib/modules/libra_system.move

```diff
@@ -0,0 +1,650 @@
+        change_events: LibraAccount::EventHandle<ValidatorSetChangeEvent>,
+    }
+
+    struct DiscoveryInfo {
+        addr: address,
+        validator_network_identity_pubkey: bytearray,
+        validator_network_address: bytearray,
+        fullnodes_network_identity_pubkey: bytearray,
+        fullnodes_network_address: bytearray,
+    }
+
+    struct DiscoverySetChangeEvent {
+        new_discovery_set: Vector::T<DiscoveryInfo>,
+    }
+
+    resource struct DiscoverySet {
+        // The current discovery set. Updated only at epoch boundaries via reconfiguration.
+        discovery_set: Vector::T<DiscoveryInfo>,
+        // Handle where discovery set change events are emitted
+        change_events: LibraAccount::EventHandle<DiscoverySetChangeEvent>,
+    }
+
+    resource struct BlockMetadata {
+      // Height of the current block
+      // TODO: Should we keep the height?
+      height: u64,
+
+      // Hash of the current block of transactions.
+      id: bytearray,
+
+      // Proposer of the current block.
+      proposer: address,
+    }
+
+    resource struct TransactionFees {
+        fee_withdrawal_capability: LibraAccount::WithdrawalCapability,
+    }
+
+    // This can only be invoked by the Association address, and only a single time.
+    // Currently, it is invoked in the genesis transaction
+    public fun initialize_validator_set() {
+      // Only callable by the validator set address
+      Transaction::assert(Transaction::sender() == 0x1D8, 1);
+
+      move_to_sender<ValidatorSet>(ValidatorSet {
+          validators: Vector::empty(),
+          additions: Vector::empty(),
+          removals: Vector::empty(),
+          change_events: LibraAccount::new_event_handle<ValidatorSetChangeEvent>(),
+      });
+    }
+
+    public fun initialize_discovery_set() {
+        // Only callable by the discovery set address
+        Transaction::assert(Transaction::sender() == 0xD15C0, 1);
+
+        move_to_sender<DiscoverySet>(DiscoverySet {
+            discovery_set: Vector::empty(),
+            change_events: LibraAccount::new_event_handle<DiscoverySetChangeEvent>(),
+        });
+    }
+
+    // This can only be invoked by the Association address, and only a single time.
+    // Currently, it is invoked in the genesis transaction
+    public fun initialize_block_metadata() {
+      // Only callable by the Association address
+      Transaction::assert(Transaction::sender() == 0xA550C18, 1);
+
+      // TODO: How should we get the default block metadata? Should it be set in the first block prologue transaction or
+      //       in the genesis?
+      move_to_sender<BlockMetadata>(BlockMetadata {
+        height: 0,
+        // FIXME: Update this once we have bytearray literals
+        id: U64Util::u64_to_bytes(0),
+        proposer: 0xA550C18,
+      });
+      LibraTimestamp::initialize_timer();
+    }
+
+    // ValidatorInfo public accessors
+
+    public fun get_validator_address(v: &ValidatorInfo): &address {
```

Comment:
Should those functions return an owned primitive value rather than a reference?

Reply:
I think they should absolutely return an owned primitive value rather than a reference. But I'm very hesitant to make these changes until we switch (or at least, in a subsequent PR that changes both the IR and Move stdlib return types for these).

Reply:
I was doing a review of the stdlib modules a while ago and wondered about the same thing because I felt a bit "scared" on returning a reference ;) 
however most of those function are later on used with equality and thus returning a reference here is somewhat natural. It's also not a mutable reference.
I just mention because as I said I went to the same process as I assume @runtian-zhou did and I am still "scared" though it seems the right thing to do....
Those function being public is a bit more of a question I think, though they are used in testing so don't know... and yet if it was about tests driving a public API that would be disgusting...